pkg/endpoint/endpoint.go

```diff
@@ -1475,6 +1477,45 @@ func (e *Endpoint) getIDandLabels() string {
 	return fmt.Sprintf("%d (%s)", e.ID, labels)
 }
 
+// MetadataResolverCB provides an implementation for resolving the endpoint
+// metadata for an endpoint such as the associated labels and annotations.
+type MetadataResolverCB func(*Endpoint) (identityLabels labels.Labels, infoLabels labels.Labels, annotations map[string]string, err error)
+
+// RunMetadataResolver starts a controller associated with the received
+// endpoint which will periodically attempt to resolve the metadata for the
+// endpoint and update the endpoint with the related. It stops resolving after
+// either the first successful metadata resolution or when the endpoint is
+// removed.
+//
+// This assumes that after the initial successful resolution, other mechanisms
+// will handle updates (such as pkg/k8s/watchers informers).
+func (e *Endpoint) RunMetadataResolver(resolveMetadata MetadataResolverCB) {
+	done := make(chan struct{})
+	controllerName := fmt.Sprintf("resolve-labels-%s", e.GetK8sNamespaceAndPodName())
+	go func() {
+		<-done
+		e.controllers.RemoveController(controllerName)
```

Comment:
Not entirely sure how much sense it makes to remove the controller here rather than just tying it to the endpoint and relying upon regular endpoint controller cleanup, but I'm not changing this functionally from the existing code. We can always clean this up, but again I was preparing this PR from a "minimal changes for a bugfix backport" prespective.